### DISCUSSION TITLE
Error in compiling dafoam-v3.0.6

### DISCUSSION THREAD

**Discussion by uttamcadambi07** [mdolab/dafoam#454] [General]
Hi,

I originally had dafoam v3.0.3 and all dependencies installed for this version from the online documentation on my Ubuntu system.

I want to install dafoam v3.0.6 side by side with v3.0.3. So I downloaded the source file, changed the Make file in $DAFOAM_ROOT/adjoint/Make/files_* so that the executables are created in $DAFOAM_ROOT/OpenFOAM/sharedLibs_v2 and similarly for sharedBins_v2. 

The original version compiled without any errors. When i source the ADR mode of OpenFOAM and compiled DAFoam, I was met with these errors 
```
In file included from DAModel/DATurbulenceModel/DAkOmegaSSTFIML.C:30:
DAModel/DATurbulenceModel/DAkOmegaSSTFIML.H:292:15: error: ‘codi::ExternalFunctionUserData’ has not been declared
  292 |         codi::ExternalFunctionUserData* d)
      |               ^~~~~~~~~~~~~~~~~~~~~~~~
DAModel/DATurbulenceModel/DAkOmegaSSTFIML.H:304:15: error: ‘codi::ExternalFunctionUserData’ has not been declared
  304 |         codi::ExternalFunctionUserData* d)
      |               ^~~~~~~~~~~~~~~~~~~~~~~~
g++ -std=c++11 -m64 -DOPENFOAM=1812 -Dlinux64 -DWM_ARCH_OPTION=64 -DWM_DP -DWM_LABEL_SIZE=32 -DCODI_AD_REVERSE -Wall -Wextra -Wno-unused-parameter -Wno-invalid-offsetof -Wno-attributes -Wno-unknown-pragmas -O3  -DNoRepository -ftemplate-depth-100 -std=c++11 -w -Wno-old-style-cast -Wno-conversion-null -Wno-deprecated-copy -DIncompressibleFlow -I/home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/TurbulenceModels/turbulenceModels/lnInclude -I/home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/TurbulenceModels/incompressible/lnInclude -I/home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/transportModels/incompressible/lnInclude -I/home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/transportModels -I/home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/transportModels/incompressible/singlePhaseTransportModel -I/home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/thermophysicalModels/radiation/lnInclude -I/home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/finiteVolume/lnInclude -I/home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/meshTools/lnInclude -I/home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/fileFormats/lnInclude -I/home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/sampling/lnInclude -I/home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/surfMesh/lnInclude -I/home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/dynamicFvMesh/lnInclude -I/home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/dynamicMesh/lnInclude -I/home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/fvOptions/lnInclude -I/home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/OpenFOAM/lnInclude -I../include -I/home/uttamcadambi/dafoam/packages/petsc-3.14.6/include -I/home/uttamcadambi/dafoam/packages/petsc-3.14.6/real-opt/include -I/home/uttamcadambi/dafoam/packages/openmpi-3.1.6/opt-gfortran/include -I/home/uttamcadambi/dafoam/packages/openmpi-3.1.6/opt-gfortran/include64 -I/home/uttamcadambi/dafoam/packages/miniconda3/include/python3.8 -I/home/uttamcadambi/dafoam/packages/miniconda3/include/python3.8  -IlnInclude -I. -I/home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/OpenFOAM/lnInclude -I/home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/OSspecific/POSIX/lnInclude -I/home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/codipack/include -I/home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/medipack/include -I/home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/medipack/src   -fPIC -c DAModel/DATurbulenceModel/DAkEpsilon.C -o Make/linux64GccDPInt32Opt/DAModel/DATurbulenceModel/DAkEpsilon.o
DAModel/DATurbulenceModel/DAkOmegaSSTFIML.C: In member function ‘void Foam::DAkOmegaSSTFIML::calcBetaField()’:
DAModel/DATurbulenceModel/DAkOmegaSSTFIML.C:925:50: error: invalid conversion from ‘void (*)(const double*, size_t, double*, size_t, int*)’ {aka ‘void (*)(const double*, long unsigned int, double*, long unsigned int, int*)’} to ‘codi::ExternalFunctionHelper<codi::ActiveReal<codi::JacobiTape<codi::JacobiTapeTypes<codi::ReverseTapeTypes<double, double, codi::LinearIndexHandler<int> >, codi::ChunkVector> > > >::PrimalFunc’ {aka ‘void (*)(const double*, long unsigned int, double*, long unsigned int, codi::DataStore*)’} [-fpermissive]
  925 |     externalFunc.callPrimalFunc(DAkOmegaSSTFIML::betaCompute);
      |                                 ~~~~~~~~~~~~~~~~~^~~~~~~~~~~
      |                                                  |
      |                                                  void (*)(const double*, size_t, double*, size_t, int*) {aka void (*)(const double*, long unsigned int, double*, long unsigned int, int*)}
In file included from /home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/codipack/include/codi.hpp:52,
                 from /home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/OpenFOAM/lnInclude/doubleScalar.H:42,
                 from /home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/OpenFOAM/lnInclude/scalar.H:40,
                 from /home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/OpenFOAM/lnInclude/token.H:42,
                 from /home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/OpenFOAM/lnInclude/runTimeSelectionTables.H:35,
                 from DAModel/DATurbulenceModel/DATurbulenceModel.H:28,
                 from DAModel/DATurbulenceModel/DAkOmegaSSTFIML.H:38,
                 from DAModel/DATurbulenceModel/DAkOmegaSSTFIML.C:30:
/home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/codipack/include/codi/tools/externalFunctionHelper.hpp:570:38: note:   initializing argument 1 of ‘void codi::ExternalFunctionHelper<CoDiType>::callPrimalFunc(codi::ExternalFunctionHelper<CoDiType>::PrimalFunc) [with CoDiType = codi::ActiveReal<codi::JacobiTape<codi::JacobiTapeTypes<codi::ReverseTapeTypes<double, double, codi::LinearIndexHandler<int> >, codi::ChunkVector> > >; codi::ExternalFunctionHelper<CoDiType>::PrimalFunc = void (*)(const double*, long unsigned int, double*, long unsigned int, codi::DataStore*)]’
  570 |       void callPrimalFunc(PrimalFunc func) {
      |                           ~~~~~~~~~~~^~~~
DAModel/DATurbulenceModel/DAkOmegaSSTFIML.C:927:24: error: ‘Tape’ is not a member of ‘codi::RealReverse’ {aka ‘codi::ActiveReal<codi::JacobiTape<codi::JacobiTapeTypes<codi::ReverseTapeTypes<double, double, codi::LinearIndexHandler<int> >, codi::ChunkVector> > >’}
  927 |     codi::RealReverse::Tape& tape = codi::RealReverse::getTape();
      |                        ^~~~
DAModel/DATurbulenceModel/DAkOmegaSSTFIML.C:927:30: error: ‘tape’ was not declared in this scope; did you mean ‘type’?
  927 |     codi::RealReverse::Tape& tape = codi::RealReverse::getTape();
      |                              ^~~~
      |                              type
DAModel/DATurbulenceModel/DAkOmegaSSTFIML.C:927:56: error: ‘getTape’ is not a member of ‘codi::RealReverse’ {aka ‘codi::ActiveReal<codi::JacobiTape<codi::JacobiTapeTypes<codi::ReverseTapeTypes<double, double, codi::LinearIndexHandler<int> >, codi::ChunkVector> > >’}
  927 |     codi::RealReverse::Tape& tape = codi::RealReverse::getTape();
      |                                                        ^~~~~~~
DAModel/DATurbulenceModel/DAkOmegaSSTFIML.C:931:49: error: invalid conversion from ‘void (*)(const double*, double*, size_t, const double*, const double*, size_t, int*)’ {aka ‘void (*)(const double*, double*, long unsigned int, const double*, const double*, long unsigned int, int*)’} to ‘codi::ExternalFunctionHelper<codi::ActiveReal<codi::JacobiTape<codi::JacobiTapeTypes<codi::ReverseTapeTypes<double, double, codi::LinearIndexHandler<int> >, codi::ChunkVector> > > >::ReverseFunc’ {aka ‘void (*)(const double*, double*, long unsigned int, const double*, const double*, long unsigned int, codi::DataStore*)’} [-fpermissive]
  931 |         externalFunc.addToTape(DAkOmegaSSTFIML::betaJacVecProd);
      |                                ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
      |                                                 |
      |                                                 void (*)(const double*, double*, size_t, const double*, const double*, size_t, int*) {aka void (*)(const double*, double*, long unsigned int, const double*, const double*, long unsigned int, int*)}
In file included from /home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/codipack/include/codi.hpp:52,
                 from /home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/OpenFOAM/lnInclude/doubleScalar.H:42,
                 from /home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/OpenFOAM/lnInclude/scalar.H:40,
                 from /home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/OpenFOAM/lnInclude/token.H:42,
                 from /home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/OpenFOAM/lnInclude/runTimeSelectionTables.H:35,
                 from DAModel/DATurbulenceModel/DATurbulenceModel.H:28,
                 from DAModel/DATurbulenceModel/DAkOmegaSSTFIML.H:38,
                 from DAModel/DATurbulenceModel/DAkOmegaSSTFIML.C:30:
/home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/src/codipack/include/codi/tools/externalFunctionHelper.hpp:603:34: note:   initializing argument 1 of ‘void codi::ExternalFunctionHelper<CoDiType>::addToTape(codi::ExternalFunctionHelper<CoDiType>::ReverseFunc, codi::ExternalFunctionHelper<CoDiType>::ForwardFunc, codi::ExternalFunctionHelper<CoDiType>::PrimalFunc) [with CoDiType = codi::ActiveReal<codi::JacobiTape<codi::JacobiTapeTypes<codi::ReverseTapeTypes<double, double, codi::LinearIndexHandler<int> >, codi::ChunkVector> > >; codi::ExternalFunctionHelper<CoDiType>::ReverseFunc = void (*)(const double*, double*, long unsigned int, const double*, const double*, long unsigned int, codi::DataStore*); codi::ExternalFunctionHelper<CoDiType>::ForwardFunc = void (*)(const double*, const double*, long unsigned int, double*, double*, long unsigned int, codi::DataStore*); codi::ExternalFunctionHelper<CoDiType>::PrimalFunc = void (*)(const double*, long unsigned int, double*, long unsigned int, codi::DataStore*)]’
  603 |       void addToTape(ReverseFunc revFunc, ForwardFunc forwFunc = nullptr, PrimalFunc primalFunc = nullptr) {
      |                      ~~~~~~~~~~~~^~~~~~~
make: *** [/home/uttamcadambi/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/wmake/rules/General/transform:35: Make/linux64GccDPInt32Opt/DAModel/DATurbulenceModel/DAkOmegaSSTFIML.o] Error 1
make: *** Waiting for unfinished jobs....
./Allmake: line 19: cd: src/pyDASolvers: No such file or directory
./Allmake: line 20: cd: src/utilities/coloring: No such file or directory
```

Seems like something to do with the data type being passed to codipack. how do i resolve this? is there a dependency update for other packages to use dafoam v3.0.6?

**Comment by uttamcadambi07** [comment]
[error.txt](https://github.com/mdolab/dafoam/files/12027015/error.txt)
Here it is in a txt file.

**Reply by uttamcadambi07** [reply]
If i skip the model that is causing the trouble in the Make file and continue the compilation, I get these errors
[error_2.txt](https://github.com/mdolab/dafoam/files/12027214/error_2.txt)

I have sourced the OpenFOAM-ADR environment before compilation.

**Comment by friedenhe** [comment]
v3.0.6 has a dependency update. So you need to recompile the ADR and ADF versions of OpenFOAM. Check: https://github.com/mdolab/dafoam/releases/tag/v3.0.6

**Reply by uttamcadambi07** [reply]
Wow these are great Dr. He. You have to definitely hold another workshop to help us understand these better. Is there one coming up? Please let us know!

**Reply by uttamcadambi07** [reply]
So i recompiled everything. and it went through without errors.
But now when i do pip install . in the dafoam (new version) directory, it detects the old version 3.0.3 and uninstalls it and installs the new version 3.0.6. But i want both versions to exist simultaneously. How can i accomplish this? Setting different pythonpaths? or virtual envs?

**Reply by uttamcadambi07** [reply]
Okay let me explain my situation a little better.

So I downloaded the file structure for dafoam but only this time, i stored it in dafoam_v2. The file structure is like OpenFOAM packages and repos. Inside dafoam_v2/repos i have dafoam-3.0.6 and inside dafoam_v2/OpenFOAM i have compiled OpenFOAM in reverse, forward and direct modes. So far so good. under dafoam_v2/repos/dafoam-3.0.6, I have managed to source direct, forward and reverse mode OpenFOAM bashrc's and compiled dafoam for all three versions. Keep in mind that all the "repos" that loadDAFoam.sh sources are from the previous installation i.e. they are located in dafoam/repos/<pyhyp,mphys,etc,etc> but the sharedBins and sharedLibs are created under dafoam_v2/OpenFOAM/sharedLibs,sharedBins

doing this, and running the script for field inversion (runScript.py from teh tutorials repo). i get this error
```
--------------------------------------------------------------------------
[[25600,1],0]: A high-performance Open MPI point-to-point messaging module
was unable to find any relevant network interfaces:

Module: OpenFabrics (openib)
  Host: uttamaero

Another transport will be used instead, although this may result in
lower performance.

NOTE: You can disable this warning by setting the MCA parameter
btl_base_warn_component_unused to 0.
--------------------------------------------------------------------------
 
-------------------------------------------------------------------------------
|                               DAFoam v3.0.6                                 |
-------------------------------------------------------------------------------
 
Traceback (most recent call last):
  File "runScript.py", line 127, in <module>
    prob.setup(mode="rev")
  File "/home/uttamcadambi/dafoam/packages/miniconda3/lib/python3.8/site-packages/openmdao/core/problem.py", line 947, in setup
    model._setup(model_comm, mode, self._metadata)
  File "/home/uttamcadambi/dafoam/packages/miniconda3/lib/python3.8/site-packages/openmdao/core/system.py", line 1455, in _setup
    self._setup_procs(self.pathname, comm, mode, self._problem_meta)
  File "/home/uttamcadambi/dafoam/packages/miniconda3/lib/python3.8/site-packages/openmdao/core/group.py", line 539, in _setup_procs
    self.setup()
  File "runScript.py", line 87, in setup
    dafoam_builder.initialize(self.comm)
  File "/home/uttamcadambi/dafoam/packages/miniconda3/lib/python3.8/site-packages/dafoam/mphys/mphys_dafoam.py", line 78, in initialize
    self.DASolver = PYDAFOAM(options=self.options, comm=comm)
  File "/home/uttamcadambi/dafoam/packages/miniconda3/lib/python3.8/site-packages/dafoam/pyDAFoam.py", line 743, in __init__
    self._initSolver()
  File "/home/uttamcadambi/dafoam/packages/miniconda3/lib/python3.8/site-packages/dafoam/pyDAFoam.py", line 2772, in _initSolver
    from .pyDASolverIncompressible import pyDASolvers
ImportError: dynamic module does not define module export function (PyInit_pyDASolverIncompressible)

```

I guess my confusions are the following:
1. Can i have the repos from the v3.0.3 installation and reuse them for the v3.0.6 installation without downloading and updating all of them? I am scared to do this because I have done a lot of implementations in dafoam 3.0.3. And by updating the packages if there is no backwards compatibility, then i am in trouble.
2. Do i need to reinstall all the repos from scratch? Including the packages such as CGNS, Ipopt, Petsc and so on?

**Reply by uttamcadambi07** [reply]
In the end, I answered my own question. So I realised that we use miniconda to segregate the site-packages that are installed and we override the local site-packages. So all I had to do was to create a new folder dafoam_v2 (as I had done before) and go through the whole process of installation from the beginning. Now I have dafoam-v3.0.3 and dafoam-v3.0.6 installed together. I could have bypassed this by possibly setting the pythonpath to the packages in the miniconda directory but did not want to mess it up because of my limited knowledge so did it this way.

**Reply by uttamcadambi07** [reply]
Hi
I am back with this again - now I am installing the update on my HPC system. 


I compiled OpenFOAM and wanted to compile ADR and ADF. When I started ADR compilation, I get this error [log.txt](https://github.com/mdolab/dafoam/files/13708877/log.txt)

Seems like it is something to do with UPStream (i used the new version of UPstream from the dafoam releases while compiling OpenFOAM). I also made all the necessary changes (I had extensively documented the hpc installation process from last time). What do you think it could be?

**Comment by friedenhe** [comment]
hmm.. the log does not really say a lot about what causes the issue. Which versions of OpenFOAM and DAFoam did you use? And which versions you are trying to update to?

**Reply by uttamcadambi07** [reply]
I am using dafoam v3.0.3 I want to update to dafoam v3.0.6. OpenFOAM v1812 in both cases

**Comment by friedenhe** [comment]
Not sure what caused this issue. However, we did update the dependencies (CoDiPack and MeDiPack) for OpenFOAM-v1812-AD in 3.0.6. We managed to compile it on our HPC without a problem..

**Reply by uttamcadambi07** [reply]
Would it help if I listed the modules and the changes that I’ve made?

**Reply by uttamcadambi07** [reply]
Update: for the version of dafoam 3.0.3, i used intel-compilers 18.0.2. But It was not working for dafoam 3.0.6 (on the HPC cluster). Now I switched to intel-compilers 19.0.3 and the ADR version of Openfoam is compiling.

Do you know what all packages have dependencies on intel-compilers that I have to recompile?

**Comment by friedenhe** [comment]
You need to update all dependencies... All things need to be compiled with the same version of intel compilers.

**Reply by uttamcadambi07** [reply]
Can you please share the versions of intel-mpi, intel-compilers, intel-mkl and python that you have used? I am not having any luck in compiling dafoam on our HPC platform. Let me give you some details. 

I am using version 3.0.6 of dafoam. So the Openfoam AD version is 1.3.0. I downloaded the latest code and compiled Openfoam in ADR and ADF using following libraries
```
  1) git/2.24.1               3) libfabric/1.7.0   5) boost/1.70.0       7) python/3.7.3           9) intel-mkl/2019.3.199
  2) intel-compilers/19.0.3   4) gcc/6.4.0         6) petsc/3.14.6.2.0   8) intel-mpi/2019.3.199

```
Please note, I had tried using intel-compilers/18.0.3 and the ADF and ADR versions would not compile. This is why I switched to 19.0.3. Additionally, I use petsc-3.14.6 which is also compiled using the same intel compilers.

I use ```mpiicc``` and ```mpiicpc``` instead of ```mpicc``` and ```mpicpc``` in $HOME/OpenFOAM-v1812-ADR(F)/wmake/rules/linux64Icc/C(c++). 

With these changes the ADR and ADF versions compile fine. dafoam complains even for the original mode for compilation. I get this error:
[dafoam_original.txt](https://github.com/mdolab/dafoam/files/13973401/dafoam_original.txt)

EDIT 1: I had to do unset WM_CODI_AD_MODE (i had forgotten to do this) and that was the reason for the error in compilation of dafoam original mode. It is now fixed. 

When i try ADR, I get this error:
[dafoam.txt](https://github.com/mdolab/dafoam/files/13973296/dafoam.txt)
So I am not sure how to go about debugging this - seems like some dependency is not sitting well with the updated codi and medipacks so it would be great if you could help us.

EDIT 2: could it be a gcc incompatibility issue? because the old version uses gcc/6.4.0 but the new version requires gcc/9.4 (we have gcc/10 series installed on our cluster and gcc/8 series but no 9 series. I tried both 10 and 8 and while 8 compiles till the end and then fails, the 10 series fails immediately)

EDIT: 3. Success!! I managed to compile it. Just for the benefit of others- you can use gcc/8.5.0 for dafoam 3.0.6. Also, python/3.7.3 works only up to pyoptsparse/2.8.1. (This is mentioned in your dependency update too).

**Reply by friedenhe** [reply]
It is great to see that it worked. Yes, you have to explore a few compiler versions. We managed to compile it with Gcc-8.5.x
